UnboundLocalError: local variable 'nla' referenced before assignment

I hit this one in production, just copy and pasting the stacktrace since pytest did such a nice job already with it.

Looks like macvlan can trigger a code path that's not properly handled?

```
primary_iface = 'eth0'

    @pytest.fixture(scope='class')
    def vlan(primary_iface):
        '''Create a macvlan device'''
        try:
>           with network.MacVLan(primary_iface) as vlan:

/usr/local/lib/python2.7/site-packages/pytest_lab/network.py:73: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <lab.network.macvlan.MacVLan object at 0x7f3a81080650>
interface = 'eth0', name = 'macvlan', dhcp = True

    def __init__(self, interface, name='macvlan', dhcp=True):
        self.name = _generate_device_name(name)
        self.dhcp = None
    
        self.ipdb = pyroute2.IPDB()
        vlan = self.ipdb.create(kind='macvlan',
                                link=self.ipdb.interfaces[interface],
                                ifname=self.name,
                                mode='bridge')
        vlan.up()
>       vlan.commit()

/usr/local/lib/python2.7/site-packages/lab/network/macvlan.py:66: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {'index': 0, 'kind': 'macvlan', 'ipaddr': (), 'ipdb_priority': 0, 'link': 4824...ridge', 'ifname': 'macvlan0', 'vlans': (), 'ports': (), 'ipdb_scope': 'create'}
tid = None
transaction = {'index': 0, 'kind': 'macvlan', 'ipaddr': (), 'ipdb_priority': 0, 'flags': 1, ...: 4824, 'ifname': 'macvlan0', 'vlans': (), 'ports': (), 'ipdb_scope': 'create'}
commit_phase = 1, commit_mask = 255, newif = True

    def commit(self,
               tid=None,
               transaction=None,
               commit_phase=1,
               commit_mask=0xff,
               newif=False):
        '''
            Commit transaction. In the case of exception all
            changes applied during commit will be reverted.
            '''
    
        if not commit_phase & commit_mask:
            return self
    
        def invalidate():
            # on failure, invalidate the interface and detach it
            # from the parent
            # 0. obtain lock on IPDB, to avoid deadlocks
            # ... all the DB updates will wait
            with self.ipdb.exclusive:
                # 1. drop the IPRoute() link
                self.nl = None
                # 2. clean up ipdb
                self.detach()
                # 3. invalidate the interface
                with self._direct_state:
                    for i in tuple(self.keys()):
                        del self[i]
                # 4. the rest
                self._mode = 'invalid'
    
        error = None
        added = None
        removed = None
        drop = True
        init = None
        debug = {'traceback': None,
                 'transaction': None,
                 'next_stage': None}
    
        if tid:
            transaction = self.global_tx[tid]
        else:
            if transaction:
                drop = False
            else:
                transaction = self.current_tx
        if transaction.partial:
            transaction.errors = []
    
        with self._write_lock:
            # if the interface does not exist, create it first ;)
            if self['ipdb_scope'] != 'system':
    
                # a special case: transition "create" -> "remove"
                if transaction['ipdb_scope'] == 'remove' and \
                        self['ipdb_scope'] == 'create':
                    self.ipdb.interfaces._detach(self['ifname'], 0)
                    return self
    
                newif = True
                self.set_target('ipdb_scope', 'system')
                try:
                    # 8<----------------------------------------------------
                    # ACHTUNG: hack for old platforms
                    if self['address'] == '00:00:00:00:00:00':
                        with self._direct_state:
                            self['address'] = None
                            self['broadcast'] = None
                    # 8<----------------------------------------------------
                    init = self.pick()
                    try:
                        request = {key: transaction[key] for key in
                                   filter(lambda x: x[:3] != 'br_',
                                          transaction)}
                        for key in ('net_ns_fd', 'net_ns_pid'):
                            if key in request:
                                with self._direct_state:
                                    self[key] = None
                                    del request[key]
>                       self.nl.link('add', **request)

/usr/local/lib/python2.7/site-packages/pyroute2/ipdb/interfaces.py:580: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyroute2.iproute.IPRoute object at 0x7f3a810800d0>, command = 16
kwarg = {'address': None, 'af_spec': None, 'bond_active_slave': None, 'bond_ad_actor_sys_prio': None, ...}
match = None, flags_dump = 769, flags_req = 5, flags_create = 1541
commands = {'add': (16, 1541), 'del': (17, 1541), 'delete': (17, 1541), 'dump': (18, 769), ...}
msg = {'index': 0, 'family': 0, '__align': 0, 'value': <class 'pyroute2.netlink.NotInitialized'>, 'header': {}, 'flags': 1, 'ifi_type': 0, 'change': 0, 'attrs': []}
lrq = <class 'pyroute2.netlink.rtnl.req.IPLinkRequest'>

    def link(self, command, **kwarg):
        '''
            Link operations.
    
            Keywords to set up ifinfmsg fields:
                * index -- interface index
                * family -- AF_BRIDGE for bridge operations, otherwise 0
                * flags -- device flags
                * change -- change mask
    
            All other keywords will be translated to NLA names, e.g.
            `mtu -> IFLA_MTU`, `af_spec -> IFLA_AF_SPEC` etc. You can
            provide a complete NLA structure or let filters do it for
            you. E.g., these pairs show equal statements::
    
                # set device MTU
                ip.link("set", index=x, mtu=1000)
                ip.link("set", index=x, IFLA_MTU=1000)
    
                # add vlan device
                ip.link("add", ifname="test", kind="dummy")
                ip.link("add", ifname="test",
                        IFLA_LINKINFO={'attrs': [['IFLA_INFO_KIND', 'dummy']]})
    
            Filters are implemented in the `pyroute2.netlink.rtnl.req` module.
            You can contribute your own if you miss shortcuts.
    
            Commands:
    
            **add**
    
            To create an interface, one should specify the interface kind::
    
                ip.link("add",
                        ifname="test",
                        kind="dummy")
    
            The kind can be any of those supported by kernel. It can be
            `dummy`, `bridge`, `bond` etc. On modern kernels one can specify
            even interface index::
    
                ip.link("add",
                        ifname="br-test",
                        kind="bridge",
                        index=2345)
    
            Specific type notes:
    
            ��� gre
    
            Create GRE tunnel::
    
                ip.link("add",
                        ifname="grex",
                        kind="gre",
                        gre_local="172.16.0.1",
                        gre_remote="172.16.0.101",
                        gre_ttl=16)
    
            The keyed GRE requires explicit iflags/oflags specification::
    
                ip.link("add",
                        ifname="grex",
                        kind="gre",
                        gre_local="172.16.0.1",
                        gre_remote="172.16.0.101",
                        gre_ttl=16,
                        gre_ikey=10,
                        gre_okey=10,
                        gre_iflags=32,
                        gre_oflags=32)
    
            Support for GRE over IPv6 is also included; use `kind=ip6gre` and
            `ip6gre_` as the prefix for its values.
    
            ��� macvlan
    
            Macvlan interfaces act like VLANs within OS. The macvlan driver
            provides an ability to add several MAC addresses on one interface,
            where every MAC address is reflected with a virtual interface in
            the system.
    
            In some setups macvlan interfaces can replace bridge interfaces,
            providing more simple and at the same time high-performance
            solution::
    
                ip.link("add",
                        ifname="mvlan0",
                        kind="macvlan",
                        link=ip.link_lookup(ifname="em1")[0],
                        macvlan_mode="private").commit()
    
            Several macvlan modes are available: "private", "vepa", "bridge",
            "passthru". Ususally the default is "vepa".
    
            ��� macvtap
    
            Almost the same as macvlan, but creates also a character tap device::
    
                ip.link("add",
                        ifname="mvtap0",
                        kind="macvtap",
                        link=ip.link_lookup(ifname="em1")[0],
                        macvtap_mode="vepa").commit()
    
            Will create a device file `"/dev/tap%s" % index`
    
            ��� tuntap
    
            Possible `tuntap` keywords:
    
                - `mode` ��� "tun" or "tap"
                - `uid` ��� integer
                - `gid` ��� integer
                - `ifr` ��� dict of tuntap flags (see ifinfmsg:... tuntap_data)
    
            Create a tap interface::
    
                ip.link("add",
                        ifname="tap0",
                        kind="tuntap",
                        mode="tap")
    
            Tun/tap interfaces are created using `ioctl()`, but the library
            provides a transparent way to manage them using netlink API.
    
            ��� veth
    
            To properly create `veth` interface, one should specify
            `peer` also, since `veth` interfaces are created in pairs::
    
                ip.link("add", ifname="v1p0", kind="veth", peer="v1p1")
    
            ��� vlan
    
            VLAN interfaces require additional parameters, `vlan_id` and
            `link`, where `link` is a master interface to create VLAN on::
    
                ip.link("add",
                        ifname="v100",
                        kind="vlan",
                        link=ip.link_lookup(ifname="eth0")[0],
                        vlan_id=100)
    
            There is a possibility to create also 802.1ad interfaces::
    
                # create external vlan 802.1ad, s-tag
                ip.link("add",
                        ifname="v100s",
                        kind="vlan",
                        link=ip.link_lookup(ifname="eth0")[0],
                        vlan_id=100,
                        vlan_protocol=0x88a8)
    
                # create internal vlan 802.1q, c-tag
                ip.link("add",
                        ifname="v100c",
                        kind="vlan",
                        link=ip.link_lookup(ifname="v100s")[0],
                        vlan_protocol=0x8100)
    
    
            ��� vrf
    
            VRF interfaces (see linux/Documentation/networking/vrf.txt)::
    
                ip.link("add",
                        ifname="vrf-foo",
                        kind="vrf",
                        vrf_table=42)
    
            ��� vxlan
    
            VXLAN interfaces are like VLAN ones, but require a bit more
            parameters::
    
                ip.link("add",
                        ifname="vx101",
                        kind="vxlan",
                        vxlan_link=ip.link_lookup(ifname="eth0")[0],
                        vxlan_id=101,
                        vxlan_group='239.1.1.1',
                        vxlan_ttl=16)
    
            All possible vxlan parameters are listed in the module
            `pyroute2.netlink.rtnl.ifinfmsg:... vxlan_data`.
    
            **set**
    
            Set interface attributes::
    
                # get interface index
                x = ip.link_lookup(ifname="eth0")[0]
                # put link down
                ip.link("set", index=x, state="down")
                # rename and set MAC addr
                ip.link("set", index=x, address="00:11:22:33:44:55", name="bala")
                # set MTU and TX queue length
                ip.link("set", index=x, mtu=1000, txqlen=2000)
                # bring link up
                ip.link("set", index=x, state="up")
    
            Keyword "state" is reserved. State can be "up" or "down",
            it is a shortcut::
    
                state="up":   flags=1, mask=1
                state="down": flags=0, mask=0
    
            **del**
    
            Destroy the interface::
    
                ip.link("del", index=ip.link_lookup(ifname="dummy0")[0])
    
            **dump**
    
            Dump info for all interfaces
    
            **get**
    
            Get specific interface info::
    
                ip.link("get", index=ip.link_lookup(ifname="br0")[0])
    
            **vlan-add**
            **vlan-del**
    
            These command names are confusing and thus are deprecated.
            Use `IPRoute.vlan_filter()`.
            '''
        if (command == 'dump') and ('match' not in kwarg):
            match = kwarg
        else:
            match = kwarg.pop('match', None)
    
        if command[:4] == 'vlan':
            log.warning('vlan filters are managed via `vlan_filter()`')
            log.warning('this compatibility hack will be removed soon')
            return self.vlan_filter(command[5:], **kwarg)
    
        flags_dump = NLM_F_REQUEST | NLM_F_DUMP
        flags_req = NLM_F_REQUEST | NLM_F_ACK
        flags_create = flags_req | NLM_F_CREATE | NLM_F_EXCL
        commands = {'set': (RTM_SETLINK, flags_create),
                    'add': (RTM_NEWLINK, flags_create),
                    'del': (RTM_DELLINK, flags_create),
                    'remove': (RTM_DELLINK, flags_create),
                    'delete': (RTM_DELLINK, flags_create),
                    'dump': (RTM_GETLINK, flags_dump),
                    'get': (RTM_GETLINK, NLM_F_REQUEST)}
    
        msg = ifinfmsg()
        # ifinfmsg fields
        #
        # ifi_family
        # ifi_type
        # ifi_index
        # ifi_flags
        # ifi_change
        #
        msg['family'] = kwarg.pop('family', 0)
        lrq = kwarg.pop('kwarg_filter', IPLinkRequest)
        (command, msg_flags) = commands.get(command, command)
        # index
        msg['index'] = kwarg.pop('index', 0)
        # flags
        flags = kwarg.pop('flags', 0) or 0
        # change
        mask = kwarg.pop('mask', 0) or kwarg.pop('change', 0) or 0
    
        # UP/DOWN shortcut
        if 'state' in kwarg:
            mask = 1                  # IFF_UP mask
            if kwarg['state'].lower() == 'up':
                flags = 1             # 0 (down) or 1 (up)
            del kwarg['state']
    
        msg['flags'] = flags
        msg['change'] = mask
    
        # apply filter
>       kwarg = lrq(kwarg)

/usr/local/lib/python2.7/site-packages/pyroute2/iproute.py:1300: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {'link': 4824, 'ipdb_priority': 0, 'IFLA_LINKINFO': {'attrs': [['IFLA_INFO_KIND', 'macvlan'], ('IFLA_INFO_DATA', {'attrs': []})]}}
argv = ({'address': None, 'af_spec': None, 'bond_active_slave': None, 'bond_ad_actor_sys_prio': None, ...},)
kwarg = {}

    def __init__(self, *argv, **kwarg):
        self.deferred = []
        self.kind = None
        self.specific = {}
        self.linkinfo = None
        self._info_data = None
>       IPRequest.__init__(self, *argv, **kwarg)

/usr/local/lib/python2.7/site-packages/pyroute2/netlink/rtnl/req.py:326: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {'link': 4824, 'ipdb_priority': 0, 'IFLA_LINKINFO': {'attrs': [['IFLA_INFO_KIND', 'macvlan'], ('IFLA_INFO_DATA', {'attrs': []})]}}
obj = {'address': None, 'af_spec': None, 'bond_active_slave': None, 'bond_ad_actor_sys_prio': None, ...}

    def __init__(self, obj=None):
        dict.__init__(self)
        if obj is not None:
>           self.update(obj)

/usr/local/lib/python2.7/site-packages/pyroute2/netlink/rtnl/req.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {'link': 4824, 'ipdb_priority': 0, 'IFLA_LINKINFO': {'attrs': [['IFLA_INFO_KIND', 'macvlan'], ('IFLA_INFO_DATA', {'attrs': []})]}}
obj = {'address': None, 'af_spec': None, 'bond_active_slave': None, 'bond_ad_actor_sys_prio': None, ...}

    def update(self, obj):
        if obj.get('family', None):
            self['family'] = obj['family']
        for key in obj:
            if key == 'family':
                continue
            v = obj[key]
            if isinstance(v, dict):
                self[key] = dict((x for x in v.items() if x[1] is not None))
            elif v is not None:
>               self[key] = v

/usr/local/lib/python2.7/site-packages/pyroute2/netlink/rtnl/req.py:38: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {'link': 4824, 'ipdb_priority': 0, 'IFLA_LINKINFO': {'attrs': [['IFLA_INFO_KIND', 'macvlan'], ('IFLA_INFO_DATA', {'attrs': []})]}}
key = 'kind', value = 'macvlan'

    def __setitem__(self, key, value):
        # ignore blacklisted attributes
        if key in self.blacklist:
            return
    
        # there must be no "None" values in the request
        if value is None:
            return
    
        # all the values must be in ascii
        try:
            if isinstance(value, unicode):
                value = value.encode('ascii')
        except NameError:
            pass
    
        if key == 'kind' and not self.kind:
            self.kind = value
>           self.flush_deferred()

/usr/local/lib/python2.7/site-packages/pyroute2/netlink/rtnl/req.py:428: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {'link': 4824, 'ipdb_priority': 0, 'IFLA_LINKINFO': {'attrs': [['IFLA_INFO_KIND', 'macvlan'], ('IFLA_INFO_DATA', {'attrs': []})]}}

    def flush_deferred(self):
        # create IFLA_LINKINFO
        linkinfo = {'attrs': []}
        self.linkinfo = linkinfo['attrs']
        dict.__setitem__(self, 'IFLA_LINKINFO', linkinfo)
        self.linkinfo.append(['IFLA_INFO_KIND', self.kind])
        # load specific NLA names
        cls = ifinfmsg.ifinfo.data_map.get(self.kind, None)
        if cls is not None:
            prefix = cls.prefix or 'IFLA_'
            for nla, _ in cls.nla_map:
                self.specific[nla] = nla
                self.specific[nla[len(prefix):].lower()] = nla
    
        # flush deferred NLAs
        for (key, value) in self.deferred:
>           if not self.set_specific(key, value):

/usr/local/lib/python2.7/site-packages/pyroute2/netlink/rtnl/req.py:354: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {'link': 4824, 'ipdb_priority': 0, 'IFLA_LINKINFO': {'attrs': [['IFLA_INFO_KIND', 'macvlan'], ('IFLA_INFO_DATA', {'attrs': []})]}}
key = 'mode', value = 'bridge'

    def set_specific(self, key, value):
        # FIXME: vlan hack
        if self.kind == 'vlan' and key == 'vlan_flags':
            if isinstance(value, (list, tuple)):
                if len(value) == 2 and \
                        all((isinstance(x, int) for x in value)):
                    value = {'flags': value[0],
                             'mask': value[1]}
                else:
                    ret = 0
                    for x in value:
                        ret |= vlan_flags.get(x, 1)
                    value = {'flags': ret,
                             'mask': ret}
            elif isinstance(value, int):
                value = {'flags': value,
                         'mask': value}
            elif isinstance(value, basestring):
                value = vlan_flags.get(value, 1)
                value = {'flags': value,
                         'mask': value}
            elif not isinstance(value, dict):
                raise ValueError()
        # the kind is known: lookup the NLA
        if key in self.specific:
            self.info_data.append((self.specific[key], value))
            return True
        elif key == 'peer' and self.kind == 'veth':
            # FIXME: veth hack
            if isinstance(value, dict):
                attrs = []
                for k, v in value.items():
                    attrs.append([ifinfmsg.name2nla(k), v])
            else:
                attrs = [['IFLA_IFNAME', value], ]
            nla = ['VETH_INFO_PEER', {'attrs': attrs}]
            self.info_data.append(nla)
            return True
        elif key == 'mode':
            # FIXME: ipvlan / tuntap / bond hack
            if self.kind == 'ipvlan':
                nla = ['IFLA_IPVLAN_MODE', value]
            elif self.kind == 'tuntap':
                nla = ['IFTUN_MODE', value]
            elif self.kind == 'bond':
                nla = ['IFLA_BOND_MODE', value]
>           self.info_data.append(nla)
E           UnboundLocalError: local variable 'nla' referenced before assignment

/usr/local/lib/python2.7/site-packages/pyroute2/netlink/rtnl/req.py:405: UnboundLocalError
```

--- a/pyroute2/netlink/rtnl/req.py
+++ b/pyroute2/netlink/rtnl/req.py
@@ -402,6 +402,8 @@ class IPLinkRequest(IPRequest):
                 nla = ['IFTUN_MODE', value]
             elif self.kind == 'bond':
                 nla = ['IFLA_BOND_MODE', value]
+            elif self.kind == 'macvlan':
+                nla = ['IFLA_MACVLAN_MODE', value]
             self.info_data.append(nla)
             return True
 

